.github/workflows/ci.yaml

```diff
@@ -0,0 +1,478 @@
+name: CI
+
+on:
+  push:
+    branches: [main]
+  merge_group:
+  pull_request:
+    branches:
+      - "main"
+      - "v[0-9]+.[0-9]+.[0-9]+"
+      - "v[0-9]+.[0-9]+"
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  lint:
+    name: Lint
+    runs-on: ubuntu-20.04 
+    steps:
+    - name: Check out code
+      uses: actions/checkout@v4
+
+    - name: Verify
+      uses: golangci/golangci-lint-action@v3
+      with:
+        version: v1.51.2
+        args: --timeout=15m0s --verbose
+  
+  unit:
+    runs-on: ubuntu-20.04
+    name: Unit and Integration tests
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+
+    - name: Setup
+      uses: ./.github/workflows/composite/setup
+
+    - name: Verify manifests
+      run: |
+        make manifests
+        git diff --exit-code
+
+    - name: Verify bin dir
+      run: |
+        BIN_FILE="metallb-operator.yaml" make bin
+        git diff --exit-code bin/
+
+    - name: Unit and Integration Tests
+      run: |
+        export KUBECONFIG=${HOME}/.kube/config
+        METALLB_BGP_TYPE=native make test
+        METALLB_BGP_TYPE=frr DEPLOY_KUBE_RBAC_PROXIES=true make test
+
+  e2e:
+    runs-on: ubuntu-20.04
+    env:
+      built_image: "metallb-operator:ci" # Arbitrary name
+    needs:
+      - lint
+      - unit
+    defaults:
+      run:
+        shell: bash
+        working-directory: metallboperator
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+
+    - name: Setup
+      uses: ./.github/workflows/composite/setup
+
+    - name: Verify manifests
+      run: |
+        make manifests
+        git diff --exit-code
+  
+    - name: Build image
+      run: |
+        IMG=${built_image} make docker-build
```

Comment:
note: this should also be in a separate lane (where we build once instead of every time), but I don't mind we do it in a different pr